File: address-geocode-international-python-examples/place_search_soap_sdk_example.py
import sys
import os

sys.path.insert(0, os.path.abspath("../address-geocode-international-python/SOAP"))

from place_search_soap import PlaceSearch
from helper import response_to_dict

def place_search_soap_sdk_go(license, is_live_key):
    print("\n" + "-" * 54)
    print("Address Geocode International - PlaceSearch - SOAP SDK")
    print("-" * 54)

    print("* Input *\n")
    print(f"SingleLine         : 17 Battery Place, New York, NY 10004")
    print(f"Address1           : ")
    print(f"Address2           : ")
    print(f"Address3           : ")
    print(f"Address4           : ")
    print(f"Address5           : ")
    print(f"Locality           : ")
    print(f"Administrative Area: ")
    print(f"Postal Code        : ")
    print(f"Country            : USA")
    print(f"Boundaries         : ")
    print(f"Max Results        : ")
    print(f"Search Type        : ")
    print(f"Extras             : ")
    print(f"Is Live            : {is_live_key}")
    print(f"License Key        : {license}")

    try:
        service = PlaceSearch(license, is_live_key, timeout_ms=10000)
        response = service.place_search(
            single_line="17 Battery Place, New York, NY 10004",
            address1="",
            address2="",
            address3="",
            address4="",
            address5="",
            locality="",
            administrative_area="",
            postal_code="",
            country="USA",
            boundaries="",
            max_results="",
            search_type="",
            extras=""
        )

        data = response_to_dict(response)
        if "Error" not in data:
            print("\n* Results *\n")
            
            print("Search Info:")
            print(f'Status           : {data["SearchInfo"].get("Status", "")}')
            print(f'NumberOfLocations: {data["SearchInfo"].get("NumberOfLocations", "")}')
            print(f'Notes            : {data["SearchInfo"].get("Notes", "")}')
            print(f'NotesDesc        : {data["SearchInfo"].get("NotesDesc", "")}')
            print(f'Warnings         : {data["SearchInfo"].get("Warnings", "")}')
            print(f'WarningDesc      : {data["SearchInfo"].get("WarningDesc", "")}')
            idx = 0;
            for key in data.keys():
                if "Locations" in key:
                    results = data[key];
                    if results is None or len(results) == 0:
                        print(f"Response Key: {key} has no records.");
                        continue;
                    idx += 1
                    print(f"Location #{idx}")
                    print(f'\tPrecisionLevel                : {data[key].get("PrecisionLevel", "")}')
                    print(f'\tType                          : {data[key].get("Type", "")}')
                    print(f'\tLatitude                      : {data[key].get("Latitude", "")}')
                    print(f'\tLongitude                     : {data[key].get("Longitude", "")}')
                    print(f'\tPremiseNumber                 : {data[key].get("PremiseNumber", "")}')
                    print(f'\tThoroughfare                  : {data[key].get("Thoroughfare", "")}')
                    print(f'\tDoubleDependentLocality       : {data[key].get("DoubleDependentLocality", "")}')
                    print(f'\tDependentLocality             : {data[key].get("DependentLocality", "")}')
                    print(f'\tLocality                      : {data[key].get("Locality", "")}')
                    print(f'\tAdministrativeArea1           : {data[key].get("AdministrativeArea1", "")}')
                    print(f'\AdministrativeArea1Abbreviation: {data[key].get("AdministrativeArea1Abbreviation", "")}')
                    print(f'\tAdministrativeArea2           : {data[key].get("AdministrativeArea2", "")}')
                    print(f'\AdministrativeArea2Abbreviation: {data[key].get("AdministrativeArea2Abbreviation", "")}')
                    print(f'\tAdministrativeArea3           : {data[key].get("AdministrativeArea3", "")}')
                    print(f'\AdministrativeArea3Abbreviation: {data[key].get("AdministrativeArea3Abbreviation", "")}')
                    print(f'\tAdministrativeArea4           : {data[key].get("AdministrativeArea4", "")}')
                    print(f'\AdministrativeArea4Abbreviation: {data[key].get("AdministrativeArea4Abbreviation", "")}')
                    print(f'\tPostalCode                    : {data[key].get("PostalCode", "")}')
                    print(f'\tCountry                       : {data[key].get("Country", "")}')
                    print(f'\tCountryISO2                   : {data[key].get("CountryISO2", "")}')
                    print(f'\tCountryISO3                   : {data[key].get("CountryISO3", "")}')
                    print(f'\tPlaceName                     : {data[key].get("PlaceName", "")}')
                    print(f'\tGoogleMapsURL                 : {data[key].get("GoogleMapsURL", "")}')
                    print(f'\tIsUnincorporated              : {data[key].get("IsUnincorporated", "")}')
                    print(f'\tStateFIPS                     : {data[key].get("StateFIPS", "")}')
                    print(f'\tCountyFIPS                    : {data[key].get("CountyFIPS", "")}')
                    print(f'\tCensusTract                   : {data[key].get("CensusTract", "")}')
                    print(f'\tCensusBlock                   : {data[key].get("CensusBlock", "")}')
                    print(f'\tCensusGeoID                   : {data[key].get("CensusGeoID", "")}')
                    print(f'\tClassFP                       : {data[key].get("ClassFP", "")}')
                    print(f'\tCongressCode                  : {data[key].get("CongressCode", "")}')
                    print(f'\tSLDUST                        : {data[key].get("SLDUST", "")}')
                    print(f'\tSLDLST                        : {data[key].get("SLDLST", "")}')
                    print(f'\tTimeZoneUTC                   : {data[key].get("TimeZone_UTC", "")}')
        else:
            print("\n* Error *\n")
            print(f'Type    : {data["Error"].get("Type", "")}')
            print(f'TypeCode: {data["Error"].get("TypeCode", "")}')
            print(f'Desc    : {data["Error"].get("Desc", "")}')
            print(f'DescCode: {data["Error"].get("DescCode", "")}')
    except Exception as ex:
        print("\nError calling service:", ex)

File: address-geocode-international-dot-net-examples/PlaceSearchSoapSdkExample.cs
using address_geocode_international_dot_net.SOAP;

namespace address_geocode_international_dot_net_examples
{
    internal static class PlaceSearchSoapSdkExample
    {
        public static void Go(string licenseKey, bool isLive)
        {
            Console.WriteLine("\r\n------------------------------------------------------");
            Console.WriteLine("Address Geocode International - PlaceSearch - SOAP SDK");
            Console.WriteLine("------------------------------------------------------");

            Console.WriteLine("\r\n* Input *\r\n");
            Console.WriteLine($"SingleLine        : 350 5th Ave, New York, NY 10118");
            Console.WriteLine($"Address1          : ");
            Console.WriteLine($"Address2          : ");
            Console.WriteLine($"Address3          : ");
            Console.WriteLine($"Address4          : ");
            Console.WriteLine($"Address5          : ");
            Console.WriteLine($"Locality          : ");
            Console.WriteLine($"AdministrativeArea: ");
            Console.WriteLine($"PostalCode        : ");
            Console.WriteLine($"Country           : US");
            Console.WriteLine($"Boundaries        : ");
            Console.WriteLine($"MaxResults        : 2");
            Console.WriteLine($"SearchType        : PostalCode");
            Console.WriteLine($"Extras            : ");
            Console.WriteLine($"Is Live           : {isLive.ToString()}");
            Console.WriteLine($"LicenseKey        : {licenseKey}");

            PlaceSearchValidation placeSearchValidation = new(isLive);

            AGIService.ResponseObject response = placeSearchValidation.PlaceSearch(
                SingleLine: "350 5th Ave, New York, NY 10118",
                Address1: "",
                Address2: "",
                Address3: "",
                Address4: "",
                Address5: "",
                Locality: "",
                AdministrativeArea: "",
                PostalCode: "",
                Country: "US",
                Boundaries: "",
                MaxResults: "2",
                SearchType: "Address",
                Extras: "",
                LicenseKey: licenseKey
             ).GetAwaiter().GetResult();

            if (!response.ContainsKey("Error"))
            {

                Console.WriteLine("\r\n* Results *\r\n");

                // Print SearchInfo essentials
                if (response.TryGetValue("SearchInfo", out var searchInfoResults) && searchInfoResults.Count > 0)
                {
                    var searchInfo = searchInfoResults[0];
                    Console.WriteLine("\r\nSearch Info:");
                    Console.WriteLine($"Status           : {(searchInfo.ContainsKey("Status") ? searchInfo["Status"] : "")}");
                    Console.WriteLine($"NumberOfLocations: {(searchInfo.ContainsKey("NumberOfLocations") ? searchInfo["NumberOfLocations"] : "")}");
                    Console.WriteLine($"Notes            : {(searchInfo.ContainsKey("Notes") ? searchInfo["Notes"] : "")}");
                    Console.WriteLine($"NotesDesc        : {(searchInfo.ContainsKey("NotesDesc") ? searchInfo["NotesDesc"] : "")}");
                    Console.WriteLine($"Warnings         : {(searchInfo.ContainsKey("Warnings") ? searchInfo["Warnings"] : "")}");
                    Console.WriteLine($"WarningDesc      : {(searchInfo.ContainsKey("WarningDesc") ? searchInfo["WarningDesc"] : "")}");

                    Console.WriteLine();
                }

                // Print locations with essential fields only
                int locationNumber = 1;
                foreach (string key in response.Keys)
                {
                    if (key.Contains("Locations"))
                    {
                        var results = response[key];
                        if (results == null || results.Count == 0)
                        {
                            Console.WriteLine($"Response Key: {key} has no records.");
                            continue;
                        }

                        foreach (var location in results)
                        {
                            Console.WriteLine("\r\nLocation #" + locationNumber++ + "\r\n");
                            Console.WriteLine("\tPrecisionLevel                 : " + (location.ContainsKey("PrecisionLevel") ? location["PrecisionLevel"] : ""));
                            Console.WriteLine("\tType                           : " + (location.ContainsKey("Type") ? location["Type"] : ""));
                            Console.WriteLine("\tLatitude                       : " + (location.ContainsKey("Latitude") ? location["Latitude"] : ""));
                            Console.WriteLine("\tLongitude                      : " + (location.ContainsKey("Longitude") ? location["Longitude"] : ""));
                            Console.WriteLine("\tPremiseNumber                  : " + (location.ContainsKey("PremiseNumber") ? location["PremiseNumber"] : ""));
                            Console.WriteLine("\tThoroughfare                   : " + (location.ContainsKey("Thoroughfare") ? location["Thoroughfare"] : ""));
                            Console.WriteLine("\tDoubleDependentLocality        : " + (location.ContainsKey("DoubleDependentLocality") ? location["DoubleDependentLocality"] : ""));
                            Console.WriteLine("\tDependentLocality              : " + (location.ContainsKey("DependentLocality") ? location["DependentLocality"] : ""));
                            Console.WriteLine("\tLocality                       : " + (location.ContainsKey("Locality") ? location["Locality"] : ""));
                            Console.WriteLine("\tAdministrativeArea1            : " + (location.ContainsKey("AdministrativeArea1") ? location["AdministrativeArea1"] : ""));
                            Console.WriteLine("\tAdministrativeArea1Abbreviation: " + (location.ContainsKey("AdministrativeArea1Abbreviation") ? location["AdministrativeArea1Abbreviation"] : ""));
                            Console.WriteLine("\tAdministrativeArea2            : " + (location.ContainsKey("AdministrativeArea2") ? location["AdministrativeArea2"] : ""));
                            Console.WriteLine("\tAdministrativeArea2Abbreviation: " + (location.ContainsKey("AdministrativeArea2Abbreviation") ? location["AdministrativeArea2Abbreviation"] : ""));
                            Console.WriteLine("\tAdministrativeArea3            : " + (location.ContainsKey("AdministrativeArea3") ? location["AdministrativeArea3"] : ""));
                            Console.WriteLine("\tAdministrativeArea3Abbreviation: " + (location.ContainsKey("AdministrativeArea3Abbreviation") ? location["AdministrativeArea3Abbreviation"] : ""));
                            Console.WriteLine("\tAdministrativeArea4            : " + (location.ContainsKey("AdministrativeArea4") ? location["AdministrativeArea4"] : ""));
                            Console.WriteLine("\tAdministrativeArea4Abbreviation: " + (location.ContainsKey("AdministrativeArea4Abbreviation") ? location["AdministrativeArea4Abbreviation"] : ""));
                            Console.WriteLine("\tPostalCode                     : " + (location.ContainsKey("PostalCode") ? location["PostalCode"] : ""));
                            Console.WriteLine("\tCountry                        : " + (location.ContainsKey("Country") ? location["Country"] : ""));
                            Console.WriteLine("\tCountryISO2                    : " + (location.ContainsKey("CountryISO2") ? location["CountryISO2"] : ""));
                            Console.WriteLine("\tCountryISO3                    : " + (location.ContainsKey("CountryISO3") ? location["CountryISO3"] : ""));
                            Console.WriteLine("\tPlaceName                      : " + (location.ContainsKey("PlaceName") ? location["PlaceName"] : ""));
                            Console.WriteLine("\tGoogleMapsURL                  : " + (location.ContainsKey("GoogleMapsURL") ? location["GoogleMapsURL"] : ""));
                            Console.WriteLine("\tIsUnincorporated               : " + (location.ContainsKey("IsUnincorporated") ? location["IsUnincorporated"] : ""));
                            Console.WriteLine("\tStateFIPS                      : " + (location.ContainsKey("StateFIPS") ? location["StateFIPS"] : ""));
                            Console.WriteLine("\tCountyFIPS                     : " + (location.ContainsKey("CountyFIPS") ? location["CountyFIPS"] : ""));
                            Console.WriteLine("\tCensusTract                    : " + (location.ContainsKey("CensusTract") ? location["CensusTract"] : ""));
                            Console.WriteLine("\tCensusBlock                    : " + (location.ContainsKey("CensusBlock") ? location["CensusBlock"] : ""));
                            Console.WriteLine("\tCensusGeoID                    : " + (location.ContainsKey("CensusGeoID") ? location["CensusGeoID"] : ""));
                            Console.WriteLine("\tClassFP                        : " + (location.ContainsKey("ClassFP") ? location["ClassFP"] : ""));
                            Console.WriteLine("\tCongressCode                   : " + (location.ContainsKey("CongressCode") ? location["CongressCode"] : ""));
                            Console.WriteLine("\tSLDUST                         : " + (location.ContainsKey("SLDUST") ? location["SLDUST"] : ""));
                            Console.WriteLine("\tSLDLST                         : " + (location.ContainsKey("SLDLST") ? location["SLDLST"] : ""));
                            Console.WriteLine("\tTimezone_UTC                   : " + (location.ContainsKey("TimeZone_UTC") ? location["TimeZone_UTC"] : ""));
                        }
                    }
                }

            }
            else
            {
                // Print only primary error fields
                if (response.TryGetValue("Error", out var errorResults) && errorResults.Count > 0)
                {
                    var error = errorResults[0];
                    Console.WriteLine("\r\n* Error *\r\n");
                    Console.WriteLine($"Type: {(error.ContainsKey("Type") ? error["Type"] : "")}");
                    Console.WriteLine($"TypeCode: {(error.ContainsKey("TypeCode") ? error["TypeCode"] : "")}");
                    Console.WriteLine($"Desc: {(error.ContainsKey("Desc") ? error["Desc"] : "")}");
                    Console.WriteLine($"DescCode: {(error.ContainsKey("DescCode") ? error["DescCode"] : "")}");
                }
                else
                {
                    Console.WriteLine("Unknown error occurred.");
                }
            }
        }
    }
}
